### ISSUE TITLE
Feature: Run AppleScript when meeting starts

### ISSUE THREAD

**Issue by shadyabhi** [leits/MeetingBar#148]
**Is your feature request related to a problem? Please describe.**
I have a huge monitor and I sometimes miss seeing the notification on the top-right about the start of the meeting. I'm looking for an alternative way of alerting myself. 

**Describe the solution you'd like**
Just like we've custom AppleScript support when joining the meeting, it would be nice to have a similar feature when the meeting starts. That way, I can probably execute an external script to have more customized notifications. I'm just talking about notifications here but the real deal with this feature request is to make it customizable for what users want to accomplish. 

**Describe alternatives you've considered**
I don't have any alternative yet.

**Comment by jgoldhammer** [comment]
Hi @shadyabhi ,

thanks for your suggestion.
Thinking about that - when exactly should the applescript be executed? Exactly on time, 3 minutes before or the same time like the notification?

Besides that do you think it is necessary to have variables available for meeting name and meeting url in the apple script runtime?

**Comment by jgoldhammer** [comment]
Some internal notes:

- Passing parameters into apple scripts via NSUserAppleScriptTask: 
https://stackoverflow.com/questions/8917050/passing-variables-to-an-applescript

- Everything about the sandbox scripting: 
https://www.objc.io/issues/14-mac/sandbox-scripting/

**Comment by shadyabhi** [comment]
@jgoldhammer  Thank you for a quick response. Since this is about "when the meeting starts", I understand that it wouldn't be possible to run the script in advance, say, "2 minutes before the meeting starts". This way, we keep the implementation easy as well? (at least version 1)

It would be nice to have the variables to make it future proof. If this feature existed today, I would have just created a dumb dialog. 

![image](https://user-images.githubusercontent.com/631797/104659139-fcaa1000-5678-11eb-9c60-6afa6e1a0008.png)

and have it centered in my screen via hammerspoon's window management API.

**Comment by jgoldhammer** [comment]
WDYT @leits ? Should we invest here?

**Comment by leits** [comment]
I think that is a very interesting idea.
I even found [how](https://stackoverflow.com/a/52544705/7540638)  to send script variables the way we run it.

The hardest aspect here is to run this action only once at the right time.
Taking into account the corner cases: change of the next meeting, a short time between the beginning of two meetings, all-day meetings.

**Comment by jgoldhammer** [comment]
I think we should think big, but start small.

- We could schedule apple script for events like the notifications with a timer. We could reuse the time option when the notification is send...
- We could not run for allday events.
- We could pass title, start date, url and maybe the event id as arguments

What a user will do is more or less on its own?

**Comment by leits** [comment]
Make sense. Would this first version meet your needs, @shadyabhi?

**Comment by shadyabhi** [comment]
@leits @jgoldhammer This sounds perfect as a start.

> We could not run for allday events.

Instead of not running for allday events, how about providing that information in the variable as well? Let's say, "meeting_type/meeting_duration". I don't know implementation details to know what would be the right variable to send. 

Assuming allday events start at 0000 hours, we can keep all events uniform, implementation wise. If users don't want AppleScript to run, they can leverage the variables to exit early.

**Comment by jgoldhammer** [comment]
The problem with allday events is that we should not trigger them multiple times. Imagine that you have to restart your mac or you restart meetingbar on the same day . In this case we do not know which allday event was already triggered or not... thinking more about that we have to persist the state in the user settings and have to cleanup after some period.

What about appointments over several days? Should we trigger every day or only the first day?

**Comment by jgoldhammer** [comment]
I have started with it and made some progress with it (handling parameter works), but I would shift this to a next release. I think, everybody is waiting for 3.0, so I will schedule this ticket for 3.1

**Comment by jgoldhammer** [comment]
You can find my first version in the attached PR. 
What do you think, @leits ?

**Comment by leits** [comment]
Will be released as part of MeetingBar 4.0.
